File: missing-subdomain-delegation-glue/python/check_delegation.py
"""Detect a missing subdomain delegation and repair it in the parent zone.

DETECT: query the parent zone's own authoritative nameserver for the NS
record set at the subdomain name, and query the child zone's own
authoritative nameserver for its NS and SOA records.

REPAIR: if the child is live but the parent has no matching NS records,
add the missing NS records in the parent zone through the Cloudflare API.

Safe to run again and again. Starts in dry run mode.

Environment:
  DNS_DOMAIN               the subdomain to check, e.g. "app.example.com"
  CLOUDFLARE_API_TOKEN     Cloudflare API token (only needed for the repair)
  CLOUDFLARE_ZONE_ID       Cloudflare zone id for the PARENT zone
  DRY_RUN                  "true" (default) reports only, "false" writes
"""
import os
import logging

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("check_delegation")


def is_delegation_missing(parent_ns_answer, child_ns_answer, child_soa_present):
    """
    parent_ns_answer: NS hostnames returned by the parent zone's authoritative
                       server when queried for the subdomain name (empty list
                       if NXDOMAIN/no data).
    child_ns_answer:  NS hostnames returned by the child zone's own
                       authoritative server for the same name.
    child_soa_present: True if the child zone answers with a valid SOA record
                        for the subdomain (i.e. the child zone is actually
                        configured and live).
    Returns True (delegation is missing/broken) when the child zone is
    live and has NS records, but the parent has no NS records for that
    name, or the parent's NS set shares no hostnames with the child's.
    """
    if not child_soa_present or not child_ns_answer:
        return False  # child isn't configured; not a delegation problem
    if not parent_ns_answer:
        return True   # parent has nothing delegating this name
    return len(set(parent_ns_answer) & set(child_ns_answer)) == 0


def _parent_zone_of(name):
    parts = name.rstrip(".").split(".")
    return ".".join(parts[1:])


def _query_ns(name, nameserver=None):
    """Query NS records for name. If nameserver is given, ask it directly."""
    import dns.resolver

    resolver = dns.resolver.Resolver()
    if nameserver:
        resolver.nameservers = [nameserver]
    try:
        answer = resolver.resolve(name, "NS")
        return sorted(str(r.target).rstrip(".") for r in answer)
    except (dns.resolver.NXDOMAIN, dns.resolver.NoAnswer, dns.resolver.NoNameservers):
        return []


def _query_soa_present(name, nameserver=None):
    import dns.resolver

    resolver = dns.resolver.Resolver()
    if nameserver:
        resolver.nameservers = [nameserver]
    try:
        resolver.resolve(name, "SOA")
        return True
    except (dns.resolver.NXDOMAIN, dns.resolver.NoAnswer, dns.resolver.NoNameservers):
        return False


def _find_authoritative_ns(zone):
    import dns.resolver

    answer = dns.resolver.resolve(zone, "NS")
    return sorted(str(r.target).rstrip(".") for r in answer)


def _add_delegation_records(subdomain, child_nameservers, api_token, zone_id, dry_run):
    """Add one NS record per child nameserver in the parent zone via Cloudflare."""
    import requests

    headers = {
        "Authorization": f"Bearer {api_token}",
        "Content-Type": "application/json",
    }
    for ns_host in child_nameservers:
        payload = {"type": "NS", "name": subdomain, "content": ns_host}
        if dry_run:
            log.info("DRY RUN: would create NS record %s -> %s", subdomain, ns_host)
            continue
        r = requests.post(
            f"https://api.cloudflare.com/client/v4/zones/{zone_id}/dns_records",
            json=payload, headers=headers, timeout=30,
        )
        r.raise_for_status()
        log.info("Created NS record %s -> %s", subdomain, ns_host)


def run():
    subdomain = os.environ["DNS_DOMAIN"]
    api_token = os.environ.get("CLOUDFLARE_API_TOKEN", "")
    zone_id = os.environ.get("CLOUDFLARE_ZONE_ID", "")
    dry_run = os.environ.get("DRY_RUN", "true").lower() == "true"

    parent_zone = _parent_zone_of(subdomain)

    parent_authoritative = _find_authoritative_ns(parent_zone)
    parent_ns_answer = []
    if parent_authoritative:
        parent_ns_answer = _query_ns(subdomain, nameserver=parent_authoritative[0])

    child_ns_answer = _query_ns(subdomain)
    child_soa_present = _query_soa_present(subdomain)

    if is_delegation_missing(parent_ns_answer, child_ns_answer, child_soa_present):
        log.warning(
            "Missing delegation for %s: parent has %s, child has %s",
            subdomain, parent_ns_answer, child_ns_answer,
        )
        if not api_token or not zone_id:
            log.info("No Cloudflare credentials set. Skipping repair, reporting only.")
            return
        _add_delegation_records(subdomain, child_ns_answer, api_token, zone_id, dry_run)
    else:
        log.info("Delegation looks fine for %s.", subdomain)


if __name__ == "__main__":
    run()

File: missing-subdomain-delegation-glue/node/check-delegation.js
/**
 * Detect a missing subdomain delegation and repair it in the parent zone.
 *
 * DETECT: query the parent zone's own authoritative nameserver for the NS
 * record set at the subdomain name, and query the child zone's own
 * authoritative nameserver for its NS and SOA records.
 *
 * REPAIR: if the child is live but the parent has no matching NS records,
 * add the missing NS records in the parent zone through the Cloudflare API.
 *
 * Safe to run again and again. Starts in dry run mode.
 *
 * Environment:
 *   DNS_DOMAIN              the subdomain to check, e.g. "app.example.com"
 *   CLOUDFLARE_API_TOKEN    Cloudflare API token (only needed for the repair)
 *   CLOUDFLARE_ZONE_ID      Cloudflare zone id for the PARENT zone
 *   DRY_RUN                 "true" (default) reports only, "false" writes
 */
import { pathToFileURL } from "node:url";

const DNS_DOMAIN = process.env.DNS_DOMAIN || "app.example.com";
const CLOUDFLARE_API_TOKEN = process.env.CLOUDFLARE_API_TOKEN || "";
const CLOUDFLARE_ZONE_ID = process.env.CLOUDFLARE_ZONE_ID || "";
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

export function isDelegationMissing(parentNsAnswer, childNsAnswer, childSoaPresent) {
  // parentNsAnswer: NS hostnames the parent zone's authoritative server
  //                 returns for the subdomain name (empty array if none).
  // childNsAnswer:  NS hostnames the child zone's own authoritative
  //                 server returns for the same name.
  // childSoaPresent: true if the child zone answers with a valid SOA
  //                  record for the subdomain (the child is actually live).
  // Returns true when the child zone is live and has NS records, but the
  // parent has no NS records for that name, or shares none with the child.
  if (!childSoaPresent || childNsAnswer.length === 0) return false;
  if (parentNsAnswer.length === 0) return true;
  const childSet = new Set(childNsAnswer);
  return !parentNsAnswer.some((ns) => childSet.has(ns));
}

function parentZoneOf(name) {
  const parts = name.replace(/\.$/, "").split(".");
  return parts.slice(1).join(".");
}

async function queryNs(name, nameserver) {
  const dns = await import("node:dns");
  const resolver = new dns.promises.Resolver();
  if (nameserver) resolver.setServers([nameserver]);
  try {
    const records = await resolver.resolveNs(name);
    return records.map((r) => r.replace(/\.$/, "")).sort();
  } catch {
    return [];
  }
}

async function querySoaPresent(name, nameserver) {
  const dns = await import("node:dns");
  const resolver = new dns.promises.Resolver();
  if (nameserver) resolver.setServers([nameserver]);
  try {
    await resolver.resolveSoa(name);
    return true;
  } catch {
    return false;
  }
}

async function findAuthoritativeNs(zone) {
  const dns = await import("node:dns");
  const records = await dns.promises.resolveNs(zone);
  return records.map((r) => r.replace(/\.$/, "")).sort();
}

async function addDelegationRecords(subdomain, childNameservers) {
  const headers = {
    Authorization: `Bearer ${CLOUDFLARE_API_TOKEN}`,
    "Content-Type": "application/json",
  };
  for (const nsHost of childNameservers) {
    const payload = { type: "NS", name: subdomain, content: nsHost };
    if (DRY_RUN) {
      console.log(`DRY RUN: would create NS record ${subdomain} -> ${nsHost}`);
      continue;
    }
    const res = await fetch(
      `https://api.cloudflare.com/client/v4/zones/${CLOUDFLARE_ZONE_ID}/dns_records`,
      { method: "POST", headers, body: JSON.stringify(payload) },
    );
    if (!res.ok) throw new Error(`Cloudflare API returned ${res.status}`);
    console.log(`Created NS record ${subdomain} -> ${nsHost}`);
  }
}

export async function run() {
  const subdomain = DNS_DOMAIN;
  const parentZone = parentZoneOf(subdomain);

  const parentAuthoritative = await findAuthoritativeNs(parentZone);
  let parentNsAnswer = [];
  if (parentAuthoritative.length > 0) {
    parentNsAnswer = await queryNs(subdomain, parentAuthoritative[0]);
  }

  const childNsAnswer = await queryNs(subdomain);
  const childSoaPresent = await querySoaPresent(subdomain);

  if (isDelegationMissing(parentNsAnswer, childNsAnswer, childSoaPresent)) {
    console.warn(
      `Missing delegation for ${subdomain}: parent has ${JSON.stringify(parentNsAnswer)}, child has ${JSON.stringify(childNsAnswer)}`,
    );
    if (!CLOUDFLARE_API_TOKEN || !CLOUDFLARE_ZONE_ID) {
      console.log("No Cloudflare credentials set. Skipping repair, reporting only.");
      return;
    }
    await addDelegationRecords(subdomain, childNsAnswer);
  } else {
    console.log(`Delegation looks fine for ${subdomain}.`);
  }
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((e) => { console.error(e); process.exit(1); });
}
